rand_distr/tests/cdf.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rand::SeedableRng;
+use rand_distr::{Distribution, Normal};
+use statrs::distribution::ContinuousCDF;
+use statrs::distribution::DiscreteCDF;
+
+/// Empirical Cumulative Distribution Function (ECDF)
+struct Ecdf {
+    sorted_samples: Vec<f64>,
+}
+
+impl Ecdf {
+    fn new(mut samples: Vec<f64>) -> Self {
+        samples.sort_by(|a, b| a.partial_cmp(b).unwrap());
+        Self {
+            sorted_samples: samples,
+        }
+    }
+
+    /// Returns the step points of the ECDF
+    /// The ECDF is a step function that increases by 1/n at each sample point
+    /// The function is continoius from the right, so we give the bigger value at the step points
+    /// First point is (-inf, 0.0), last point is (max(samples), 1.0)
+    fn step_points(&self) -> Vec<(f64, f64)> {
+        let mut points = Vec::with_capacity(self.sorted_samples.len() + 1);
+        let mut last = f64::NEG_INFINITY;
+        let mut count = 0;
+        let n = self.sorted_samples.len() as f64;
+        for &x in &self.sorted_samples {
+            if x != last {
+                points.push((last, count as f64 / n));
+                last = x;
+            }
+            count += 1;
+        }
+        points.push((last, count as f64 / n));
+        points
+    }
+}
+
+fn kolmo_smirnov_statistic_continuous(ecdf: Ecdf, cdf: impl Fn(f64) -> f64) -> f64 {
+    let mut max_diff = 0.;
+    // The maximum will always be at a step point, because the cdf is continious monotonic increasing
+    let step_points = ecdf.step_points();
+    for i in 0..step_points.len() - 1 {
+        // This shift is because we want the value of the ecdf at (x_{i + 1} - epsilon) = ecdf(x_i) and comare it to cdf(x_{i + 1})
+        // cdf(x_{i + 1} - epsilon) = cdf(x_{i+1}) because cdf is continious
+        let x = step_points[i + 1].0;
+        let diff = (step_points[i].1 - cdf(x)).abs();
+
+        if diff > max_diff {
+            max_diff = diff;
+        }
+    }
+    max_diff
+}
+
+fn kolmo_smirnov_statistic_discrete(ecdf: Ecdf, cdf: impl Fn(i64) -> f64) -> f64 {
```

Comment:
statrs calls this a discrete CDF (true), but more specifically (and importantly) it is only defined on the domain of integers...

... but then there's a bug here: your `step_points` method yields the top/left points of the CDF, but there may be additional positions to the right with the same y-coordinate. I.e. there may (probably) exist `i` such that `step_points[i+1].0 - 1 > step_points[i].0`.

And (not being familiar with the domain), I'm still not certain whether this approach to KS is valid on a discrete domain. WP gives [an alternative approach](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test#Discrete_and_mixed_null_distribution) using the inverse CDF, and mentions some further approaches (see the KSgeneral R package).

Reply:
You are correct the calculation can be a bit incorrect (which big sample sizes it should not be too bad), I will rethink what would be best for distributions with integer support.

Reply:
In the literature I find that the test is too conservative if the distribution is discrete.
In our case this would mean, the critical value is bigger than it could be. We have a higher change of not detecting a deviation.
I do not thing it is a problem, because this should get better with increased sample sizes and compared to applications of statistical tests, we have the luxury of almost unlimited amounts of samples.